### PR TITLE
applications: nrf_desktop: QoS service

### PR DESCRIPTION
**Pull request by pdunaj**
Jira:DESK-1024

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

File: applications/nrf_desktop/src/modules/qos.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+#include "chmap_filter.h"
+
+#include "ble_event.h"
+
+#define MODULE qos
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_BAS_LOG_LEVEL);
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/qos.c:21]: QOS_LOG_LEVEL

File: applications/nrf_desktop/src/modules/qos.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		0x4c, 0x65, 0xe6, 0xe6, 0x98, 0xb6, 0x78, 0xd5,
+		0x7d, 0x17, 0x19, 0x46, 0x75, 0x2, 0x47, 0x24);
+
+static bool active;
+static u8_t chmap[CHMAP_BLE_BITMASK_SIZE];
+static struct k_spinlock lock;
+
+static void chmap_ccc_cfg_changed(const struct bt_gatt_attr *attr,
+				 u16_t value)
+{
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/qos.c:37]: We should use spinlock for active too

File: applications/nrf_desktop/src/modules/qos.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	active = (value == BT_GATT_CCC_NOTIFY);
+}
+
+static ssize_t read_chmap(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			  void *buf, u16_t len, u16_t offset)
+{
+	const void *value = attr->user_data;
+	__ASSERT_NO_MSG(value = &chmap);
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/qos.c:45]: ==

File: applications/nrf_desktop/src/modules/qos.c (hunk 2 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/qos.c:37]: Or make active as atomic

File: applications/nrf_desktop/src/modules/qos.c [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	memcpy(chmap_local, chmap, sizeof(chmap));
+	k_spin_unlock(&lock, key);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, chmap_local,
+				 sizeof(chmap_local));
+}
+
+/* Battery Service Declaration */
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/qos.c:57]: QoS

File: applications/nrf_desktop/src/modules/qos.c [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_BAS_LOG_LEVEL);
+
+
+static struct bt_uuid_128 custom_desc_uuid = BT_UUID_INIT_128(
+		0x4c, 0x65, 0xe6, 0xe6, 0x98, 0xb6, 0x78, 0xd5,
+		0x7d, 0x17, 0x19, 0x46, 0x75, 0x2, 0x47, 0x24);
+static const struct bt_uuid_128 custom_desc_chrc_uuid = BT_UUID_INIT_128(
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/qos.c:27]: Can we use the same UUID for service and characteristic ?

File: applications/nrf_desktop/src/modules/qos.c (hunk 2 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/qos.c:37]: we can to be formally correct but it does not matter on our arch
bas has the same problem btw

File: applications/nrf_desktop/src/modules/qos.c (hunk 5 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/qos.c:27]: it should work but as discussed offline I will use different values

**Pull request merged**